### PR TITLE
expression: remove the usage of "TypeClass" in "builtin_cast.go"

### PR DESCRIPTION
**Pull request by zz-jason**
to #4394

### CODE REVIEW

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -122,31 +122,31 @@ func (c *castAsIntFunctionClass) getFunction(ctx context.Context, args []Express
...
+		sig = &builtinCastDurationAsIntSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_CastDurationAsInt)
+	case tpJSON:
+		sig = &builtinCastJSONAsIntSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_CastJsonAsInt)
+	case tpString:
+		sig = &builtinCastStringAsIntSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_CastStringAsInt)
+	default:
+		panic("unsupported evalTp in castAsIntFunctionClass")
```
Comment (zimulala) [inline expression/builtin_cast.go]: Return an error is better.

Comment (zz-jason) [reply expression/builtin_cast.go]: no, we can discover programmer's mistake through this panic.

**Comment by zimulala** [comment]
LGTM

**Comment by zz-jason** [comment]
do not merge now, wait for https://github.com/pingcap/tidb/pull/4543

**Pull request merged**